src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateDeconstructMethodTests.cs

```diff
@@ -0,0 +1,241 @@
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out object x, out object y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        (var x, var y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionDeclaration_UndeterminedType2()
+        {
+            await TestInRegularAndScript1Async(
+@"class Class
+{
+    void Method()
+    {
+        var (x, y) = [|this|];
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out object x, out object y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        var (x, y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionDeclaration_BuiltinType()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        (int x, int y) = [|1|];
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        int x, y;
+        (x, y) = [|this|];
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out int x, out int y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        int x, y;
+        (x, y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment_Nested()
+        {
+            // We only offer a fix for non-nested deconstruction, at the moment
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        int x, y, z;
+        ((x, y), z) = ([|this|], 0);
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment_Array()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        int x, y;
+        (x, y) = [|new[] { this }|];
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestSimpleDeconstructionForeach()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        foreach ((int x, int y) in new[] { [|this|] }) { }
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out int x, out int y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        foreach ((int x, int y) in new[] { this }) { }
+    }
+}");
```

Comment:
can we have a cross-type generation, so we can verify the accessibility of the generated method is correct?

Reply:
can we have some decosntructors that use types that include type parameters (both type-symbol and method-symbol type parameters)?